File: app/src/main/java/de/bitb/astroskop/ui/main/circumstances/CircumstancesPresenter.java
package de.bitb.astroskop.ui.main.circumstances;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import de.bitb.astroskop.AstroApplication;
import de.bitb.astroskop.controller.CircumstanceController;
import de.bitb.astroskop.controller.ControllerCallback;
import de.bitb.astroskop.injection.IInjection;
import de.bitb.astroskop.injection.components.AppComponent;
import de.bitb.astroskop.model.Circumstance;
import de.bitb.astroskop.ui.base.BasePresenter;
import de.bitb.astroskop.ui.base.adapter.AdapterPresenter;
import de.bitb.astroskop.ui.base.adapter.IAdapterView;

public class CircumstancesPresenter extends BasePresenter<ICircumstancesView> implements AdapterPresenter<Circumstance>, IInjection {

    @Inject
    protected CircumstanceController circumstanceController;

    private List<Circumstance> circumstances = new ArrayList<>();

    public CircumstancesPresenter(AstroApplication application, ICircumstancesView view) {
        super(application, view);
    }

    @Override
    public void inject(AppComponent appComponent) {
        appComponent.inject(this);
    }

    public void onCreate() {
        refreshView();
    }

    public void onGoClick() {
        circumstanceController.createRandomCircumstance();
        refreshView();
    }

    private void refreshView() {
        circumstances = circumstanceController.getAllCircumstances();
        getView().refreshView();
    }

    @Override
    public int getAdapterItemCount() {
        return circumstances.size();
    }

    @Override
    public void onBindAtPosition(IAdapterView<Circumstance> holder, int position) {
        holder.bindValues(circumstances.get(position));
    }

    @Override
    public void onItemClicked(Circumstance model) {
        getView().openCircumstanceDetails(model);
    }

    @Override
    public void onItemLongClicked(Circumstance circumstance) {
        getView().openDeleteDialog(circumstance);
    }

    public void deleteCircumstance(Circumstance circumstance) {
        circumstanceController.delete(circumstance);
        refreshView();
    }

}

File: app/src/main/java/de/bitb/astroskop/ui/main/circumstances/CircumstancesPresenter.kt
package de.bitb.astroskop.ui.main.circumstances

import java.util.ArrayList

import javax.inject.Inject

import de.bitb.astroskop.BaseApplication
import de.bitb.astroskop.controller.CircumstanceController
import de.bitb.astroskop.injection.IInjection
import de.bitb.astroskop.injection.components.AppComponent
import de.bitb.astroskop.model.Circumstance
import de.bitb.astroskop.ui.base.BasePresenter
import de.bitb.astroskop.ui.base.adapter.AdapterPresenter
import de.bitb.astroskop.ui.base.adapter.IAdapterView

class CircumstancesPresenter(application: BaseApplication, view: ICircumstancesView) : BasePresenter<ICircumstancesView>(application, view), AdapterPresenter<Circumstance>, IInjection {

    @Inject
    var circumstanceController: CircumstanceController? = null

    private var circumstances: List<Circumstance> = ArrayList()

    override val adapterItemCount: Int
        get() = circumstances.size

    override fun inject(appComponent: AppComponent?) {
        appComponent?.inject(this)
    }

    fun onCreate() {
        refreshView()
    }

    fun onGoClick() {
        circumstanceController!!.createRandomCircumstance()
        refreshView()
    }

    private fun refreshView() {
        circumstances = circumstanceController!!.allCircumstances
        getView().refreshView()
    }

    override fun onBindAtPosition(holder: IAdapterView<Circumstance>, position: Int) {
        holder.bindValues(circumstances[position])
    }

    override fun onItemClicked(model: Circumstance) {
        getView().openCircumstanceDetails(model)
    }

    override fun onItemLongClicked(circumstance: Circumstance) {
        getView().openDeleteDialog(circumstance)
    }

    fun deleteCircumstance(circumstance: Circumstance) {
        circumstanceController!!.delete(circumstance)
        refreshView()
    }

}
